### PR TITLE
added timeout to anger

### LINKED ISSUE #11: Per request timeout
I would like to set a timeout option that applies to each sender request, within which if all expected responses haven't been received we continue to the next trigger or complete if all done.

The existing `tail` option kind of does this if there's no more requests to be sent:

```
tail: 200
requests: 1
senders: 1
connections: 3

----------x                        0ms (request sent)
          -------x                 +80ms receive response                                   
          -----------x             +130ms receive response
          -------------------!     +200ms tail is triggered, 1 missing response
```
However if you have >= 2 requests and 1 sender it never gets triggered, so it hangs forever:

```
tail: 200
requests: 2
senders: 1
connections: 3

----------x                                 0ms (request sent)
        -------x                            +80ms receive response                                   
        -----------x                        +130ms receive response
        -------------------!----Infinity    No timeout, we wait forever
```

In the case above, I'd like to timeout after 200ms, mark the response as missing and start the next request.

```
timeout: 200
requests: 2
senders: 1
connections: 3

----------x                                 0ms (request sent)
        -------x                            +80ms receive response                                   
        -----------x                        +130ms receive response
        -------------------!                Timeout, 1 missing response
                           x                Next request triggered...
```

### PR DESCRIPTION
**Pull request by thekemkid**
Fixes #11

### CODE REVIEW

File: anger.js [hunk 1]
```diff
@@ -67,18 +69,36 @@ function anger (opts) {
...
-    const end = process.hrtime(startTime)
-    const responseTime = end[0] * 1e3 + end[1] / 1e6
-    latencies.record(responseTime)
+    recordResponseTime(mapObj.start)
 
     if (!--mapObj.expectedResponses) {
-      tracker.emit('publish-events-recieved', uidOf(payload))
+      mapObj.finished = true
+      if (timeout) clearTimeout(mapObj.timeout)
+      tracker.emit('publish-events-recieved', uid)
```
Comment (mcollina) [inline anger.js]: typo

Comment (davidmarkclements) [inline anger.js]: i before e except after c :neckbeard:

File: anger.js [hunk 2]
```diff
@@ -103,7 +123,16 @@ function anger (opts) {
   function triggerSender (sender) {
     const uid = opts.trigger(sender)
     if (++totalRequests === opts.requests && tail) setTimeout(complete, tail)
-    map.set(uid, { start: process.hrtime(), sender: sender, expectedResponses: clients.length, id: uid })
+    let replyTimeout
+    if (timeout) replyTimeout = setTimeout(handleTimeout(uid), timeout)
```
Comment (davidmarkclements) [inline anger.js]: I don't think standard style allows for no braces around if, 

do we not have linting set up for this project?

Comment (mcollina) [inline anger.js]: we do: https://github.com/nearform/anger/blob/master/package.json#L7.

Comment (thekemkid) [inline anger.js]: So does this need a fix or..? 😅

**Comment by mcollina** [comment]
LGTM

**Comment by mcollina** [comment]
@mtharrison ok for you?

**Pull request merged**